packages/react-native/ReactCommon/react/nativemodule/core/platform/ios/ReactCommon/RCTTurboModule.mm

```diff
@@ -359,13 +400,52 @@ id convertJSIValueToObjCObject(jsi::Runtime &runtime, const jsi::Value &value, s
           rejectWasCalled = YES;
         };
 
-        invokeCopy(resolveBlock, rejectBlock);
+        RCTPromiseRejectBlock rejectBlock = ^(NSString *code, NSString *message, NSError *error) {
+          NSDictionary *reason = RCTJSErrorFromCodeMessageAndNSError(code, message, error);
+          handlePromiseRejection(reason);
+        };
+
+        RCTNSDictionaryPromiseRejectBlock internalRejectBlock = nil;
+
+        if (RCTRejectTurboModulePromiseOnNativeError()) {
+          internalRejectBlock = ^(NSDictionary *dict) {
+            handlePromiseRejection(dict);
+          };
+        }
+
+        invokeCopy(resolveBlock, rejectBlock, internalRejectBlock);
         return jsi::Value::undefined();
       });
 
   return Promise.callAsConstructor(runtime, fn);
 }
 
+static std::optional<jsi::JSError> handleCause(
+    jsi::Runtime &runtime,
+    bool isSync,
+    NSDictionary *cause,
+    NSMutableArray *retainedObjectsForInvocation,
+    RCTNSDictionaryPromiseRejectBlock optionalInternalRejectBlock)
+{
+  if (isSync) {
+    NSString *message = cause[@"message"];
+    if (message) {
+      return convertNSDictionaryToJSError(runtime, std::string([message UTF8String]), cause);
+    } else {
+      return convertNSDictionaryToJSError(runtime, "<unknown>", cause);
+    }
+  } else {
+    if (optionalInternalRejectBlock != nil) {
+      optionalInternalRejectBlock(cause);
+    } else {
+      // Crash on native layer there is no promise in JS to reject.
+      // We executed JSFunction returning void asynchrounously.
+      throw;
+    }
+  }
+  return std::nullopt;
```

Comment:
Move this after `optionalInternalRejectBlock` so it's clear that's the only case we can return null.